File: web/src/main/java/com/walter/lifelog/web/config/KafkaConsumerConfig.java
package com.walter.lifelog.web.config;

import com.fasterxml.jackson.databind.deser.std.StringDeserializer;
import com.walter.lifelog.shared.util.KafkaConnectHelper;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.io.Resource;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.core.ConsumerFactory;
import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
import org.springframework.kafka.support.serializer.JacksonJsonDeserializer;

import java.util.HashMap;

@Configuration
public class KafkaConsumerConfig {
    @Value("${spring.kafka.bootstrap-servers:}")
    private String bootstrapServers;

    @Value("${kafka.ssl.key-location:}")
    private Resource sslKeyResource;

    @Value("${kafka.ssl.certificate-location:}")
    private Resource sslCertResource;

    @Value("${kafka.ssl.ca-location:}")
    private Resource sslCaResource;

    @Value("${spring.kafka.consumer.group-id:lifelog-web-group}")
    private String consumerGroupId;

    @Bean
    public ConsumerFactory<String, Object> consumerFactory() {
        final HashMap<String, Object> properties = KafkaConnectHelper.getSslProperties(sslKeyResource, sslCertResource, sslCaResource);
        properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        properties.put(ConsumerConfig.GROUP_ID_CONFIG, consumerGroupId);
        properties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        properties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, JacksonJsonDeserializer.class);
        properties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
        properties.put(JacksonJsonDeserializer.TRUSTED_PACKAGES, "com.walter.lifelog.*");
        return new DefaultKafkaConsumerFactory<>(properties);
    }

    @Bean
    public ConcurrentKafkaListenerContainerFactory<String, Object> kafkaListenerContainerFactory(ConsumerFactory<String, Object> consumerFactory) {
        final ConcurrentKafkaListenerContainerFactory<String, Object> factory = new ConcurrentKafkaListenerContainerFactory<>();
        factory.setConsumerFactory(consumerFactory);
        factory.setConcurrency(3);
        return factory;
    }
}

File: worker/src/main/kotlin/com/walter/lifelog/worker/config/KafkaConsumerConfig.kt
package com.walter.lifelog.worker.config

import com.walter.lifelog.shared.util.KafkaConnectHelper
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.common.serialization.StringDeserializer
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.io.Resource
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory
import org.springframework.kafka.core.ConsumerFactory
import org.springframework.kafka.core.DefaultKafkaConsumerFactory
import org.springframework.kafka.support.serializer.JacksonJsonDeserializer

@Configuration
class KafkaConsumerConfig {
    companion object {
        private val MAX_CONCURRENCY = Runtime.getRuntime().availableProcessors()
    }

    @Value("\${spring.kafka.bootstrap-servers:}")
    private lateinit var bootstrapServers: String

    @Value("\${kafka.ssl.key-location}")
    private lateinit var sslKeyResource: Resource

    @Value("\${kafka.ssl.certificate-location}")
    private lateinit var sslCertResource: Resource

    @Value("\${kafka.ssl.ca-location}")
    private lateinit var sslCaResource: Resource

    @Value("\${spring.kafka.consumer.group-id:lifelog-worker-group}")
    private lateinit var consumerGroupId: String

    @Bean
    fun consumerFactory(): ConsumerFactory<String, Any> {
        val properties = KafkaConnectHelper.getSslProperties(sslKeyResource, sslCertResource, sslCaResource)
        properties[ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG] = bootstrapServers
        properties[ConsumerConfig.GROUP_ID_CONFIG] = consumerGroupId
        properties[ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG] = StringDeserializer::class.java
        properties[ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG] = JacksonJsonDeserializer::class.java
        properties[ConsumerConfig.AUTO_OFFSET_RESET_CONFIG] = "earliest"
        properties[JacksonJsonDeserializer.TRUSTED_PACKAGES] = "com.walter.lifelog.*"
        return DefaultKafkaConsumerFactory(properties)
    }

    @Bean
    fun kafkaListenerContainerFactory(consumerFactory: ConsumerFactory<String, Any>): ConcurrentKafkaListenerContainerFactory<String, Any> {
        val factory = ConcurrentKafkaListenerContainerFactory<String, Any>()
        factory.setConsumerFactory(consumerFactory)
        factory.setConcurrency(MAX_CONCURRENCY)
        return factory
    }
}